// @microsoft/generator-sharepoint/lib/generators/applicationCustomizer/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApplicationCustomizerGenerator = exports.defineOptions = exports.composeWith = void 0;
const path = __importStar(require("path"));
const Utils = __importStar(require("../../common/utilities"));
const BaseExtension = __importStar(require("../extension/BaseExtensionGenerator"));
function composeWith(base, options) {
    Utils.compose(path.basename(__dirname), base, options);
}
exports.composeWith = composeWith;
function defineOptions(generator) {
    BaseExtension.defineOptions(generator, 'Application Customizer');
}
exports.defineOptions = defineOptions;
class ApplicationCustomizerGenerator extends BaseExtension.BaseExtensionGenerator {
    constructor(args, options) {
        super(args, options);
        this.friendlyName = 'Application Customizer';
        this.codeName = 'ApplicationCustomizer';
        this.allowedTemplates = [];
        defineOptions(this);
    }
    initializing() {
        /* no-op */
    }
    async prompting() {
        this.ensureCorrectFolder();
        if (this.shouldExecute()) {
            await super.prompting();
        }
    }
    configuring() {
        if (this.shouldExecute()) {
            super.configuring();
            this.ensureDependencyGroup('applicationCustomizer');
            this.serveJson.addConfiguration(this.context.componentNameCamelCase, {
                pageUrl: 'https://{tenantDomain}/SitePages/myPage.aspx',
                customActions: {
                    [this.context.componentId]: {
                        location: 'ClientSideExtension.ApplicationCustomizer',
                        properties: {
                            testMessage: 'Test message'
                        }
                    }
                }
            });
        }
    }
    writing() {
        if (this.shouldExecute()) {
            super.writing();
        }
    }
    install() {
        if (this.shouldExecute()) {
            super.install();
        }
    }
    end() {
        super.end();
    }
    includeClientSideInstances() {
        return true;
    }
    shouldExecute() {
        return (this.config.get('environment') === 'spo' &&
            this.config.get('componentType') === 'extension' &&
            this.config.get('extensionType') === 'ApplicationCustomizer');
    }
}
exports.ApplicationCustomizerGenerator = ApplicationCustomizerGenerator;
// Yeoman expects exports of this type, which overrides the TS exports above
// Ensure that anything marked as export above is also listed here
module.exports = ApplicationCustomizerGenerator;
module.exports.defineOptions = defineOptions;
module.exports.composeWith = composeWith;
