// @microsoft/generator-sharepoint/lib/generators/adaptiveCardExtension/templates/none/{componentClassName}.ts
import { IPropertyPaneConfiguration } from '@microsoft/sp-property-pane';
import { BaseAdaptiveCardExtension } from '@microsoft/sp-adaptive-card-extension-base';
import { CardView } from './cardView/CardView';
import { QuickView } from './quickView/QuickView';
import { <%= componentName %>PropertyPane } from './<%= componentName %>PropertyPane';

export interface I<%= componentClassName %>Props {
  title: string;
}

export interface I<%= componentClassName %>State {
}

const CARD_VIEW_REGISTRY_ID: string = '<%= componentName %>_CARD_VIEW';
export const QUICK_VIEW_REGISTRY_ID: string = '<%= componentName %>_QUICK_VIEW';

export default class <%= componentClassName %> extends BaseAdaptiveCardExtension<
  I<%= componentClassName %>Props,
  I<%= componentClassName %>State
> {
  private _deferredPropertyPane: <%= componentName %>PropertyPane | undefined;

  public onInit(): Promise<void> {
    this.state = { };

    this.cardNavigator.register(CARD_VIEW_REGISTRY_ID, () => new CardView());
    this.quickViewNavigator.register(QUICK_VIEW_REGISTRY_ID, () => new QuickView());

    return Promise.resolve();
  }

  protected loadPropertyPaneResources(): Promise<void> {
    return import(
      /* webpackChunkName: '<%= componentName %>-property-pane'*/
      './<%= componentName %>PropertyPane'
    )
      .then(
        (component) => {
          this._deferredPropertyPane = new component.<%= componentName %>PropertyPane();
        }
      );
  }

  protected renderCard(): string | undefined {
    return CARD_VIEW_REGISTRY_ID;
  }

  protected getPropertyPaneConfiguration(): IPropertyPaneConfiguration {
    return this._deferredPropertyPane?.getPropertyPaneConfiguration();
  }
}
