// @microsoft/generator-sharepoint/lib/generators/library/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LibraryGenerator = exports.defineOptions = exports.composeWith = void 0;
const path = __importStar(require("path"));
const Utils = __importStar(require("../../common/utilities"));
const BaseComponent = __importStar(require("../component/BaseComponentGenerator"));
function composeWith(base, options) {
    Utils.compose(path.basename(__dirname), base, options);
}
exports.composeWith = composeWith;
function defineOptions(generator) {
    BaseComponent.defineOptions(generator, 'Library');
}
exports.defineOptions = defineOptions;
class LibraryGenerator extends BaseComponent.BaseComponentGenerator {
    constructor(args, options) {
        super(args, options);
        this.friendlyName = 'Library';
        this.codeName = 'Library';
        this.allowedTemplates = [];
        this.folderName = 'libraries';
        defineOptions(this);
    }
    initializing() {
        /* no-op */
    }
    async prompting() {
        this.ensureCorrectFolder();
        if (this.shouldExecute()) {
            await super.prompting();
        }
    }
    configuring() {
        if (this.shouldExecute()) {
            const bundleEntry = {
                components: [
                    {
                        entrypoint: `./lib/index.js`,
                        manifest: `./src/libraries/${this.context.componentNameCamelCase}/${this.context.componentClassName}.manifest.json`
                    }
                ]
            };
            super.configuring(bundleEntry);
            this.ensureDependencyGroup('library');
            this.packageJsonManager.data.private = false;
        }
    }
    writing() {
        if (this.shouldExecute()) {
            super.writing(true);
            this.copyTemplate(this.templatePath('./index/'), this.destinationPath('./src'));
        }
    }
    install() {
        if (this.shouldExecute()) {
            super.install();
        }
    }
    end() {
        super.end();
    }
    shouldExecute() {
        return this.config.get('environment') === 'spo' && this.config.get('componentType') === 'library';
    }
}
exports.LibraryGenerator = LibraryGenerator;
// Yeoman expects exports of this type, which overrides the TS exports above
// Ensure that anything marked as export above is also listed here
module.exports = LibraryGenerator;
module.exports.defineOptions = defineOptions;
module.exports.composeWith = composeWith;
