// @microsoft/generator-sharepoint/lib/generators/extension/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExtensionSelectorGenerator = exports.composeWith = exports.defineOptions = void 0;
const colors_1 = __importDefault(require("colors"));
const path = __importStar(require("path"));
const BaseGenerator_1 = require("../../common/BaseGenerator");
const Utils = __importStar(require("../../common/utilities"));
const CommandSet = __importStar(require("../commandSet"));
const FieldCustomizer = __importStar(require("../fieldCustomizer"));
const FormCustomizer = __importStar(require("../formCustomizer"));
function defineOptions(generator) {
    FieldCustomizer.defineOptions(generator);
    CommandSet.defineOptions(generator);
    FormCustomizer.defineOptions(generator);
    generator.option('extension-type', {
        description: `The type of extension. Currently "ApplicationCustomizer", "FieldCustomizer", "ListViewCommandSet", "FormCustomizer", or "SearchQueryModifier"`,
        type: String
    });
}
exports.defineOptions = defineOptions;
function composeWith(base, options) {
    Utils.compose(path.basename(__dirname), base, options);
}
exports.composeWith = composeWith;
/**
 * This class selects between multiple different types of extensions for instantiation.
 */
class ExtensionSelectorGenerator extends BaseGenerator_1.BaseGenerator {
    constructor(args, options) {
        super(args, options);
        this.friendlyName = 'Extension Selection';
    }
    initializing() {
        /* no-op */
    }
    async prompting() {
        this.ensureCorrectFolder();
        if (!this.shouldExecute()) {
            return;
        }
        const choices = [
            {
                name: 'Application Customizer',
                value: 'ApplicationCustomizer'
            },
            {
                name: 'Field Customizer',
                value: 'FieldCustomizer'
            },
            {
                name: 'ListView Command Set',
                value: 'ListViewCommandSet'
            },
            {
                name: 'Form Customizer',
                value: 'FormCustomizer'
            },
            {
                name: 'Search Query Modifier',
                value: 'SearchQueryModifier'
            }
        ];
        const validExtensionTypes = new Set(choices.map((choice) => choice.value));
        const answers = await this.prompt([
            {
                type: 'list',
                name: 'extensionType',
                default: 'ApplicationCustomizer',
                when: () => !this.config.get('extensionType'),
                message: 'Which type of client-side extension to create?',
                choices
            }
        ]);
        const extensionType = answers.extensionType || this.config.get('extensionType');
        if (validExtensionTypes.has(extensionType)) {
            this.config.set('extensionType', extensionType);
        }
        else {
            this.log(colors_1.default.red(`Invalid extensionType: "${extensionType}"`));
        }
    }
    configuring() {
        /* no-op */
    }
    writing() {
        /* no-op */
    }
    install() {
        /* no-op */
    }
    end() {
        /* no-op */
    }
    shouldExecute() {
        return this.config.get('environment') === 'spo' && this.config.get('componentType') === 'extension';
    }
}
exports.ExtensionSelectorGenerator = ExtensionSelectorGenerator;
// Yeoman expects exports of this type, which overrides the TS exports above
// Ensure that anything marked as export above is also listed here
module.exports = ExtensionSelectorGenerator;
module.exports.defineOptions = defineOptions;
module.exports.composeWith = composeWith;
