// @microsoft/generator-sharepoint/lib/generators/webpart/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebpartGenerator = exports.defineOptions = exports.composeWith = void 0;
const path = __importStar(require("path"));
const Utils = __importStar(require("../../common/utilities"));
const YeomanConfiguration_1 = require("../../common/YeomanConfiguration");
const BaseComponent = __importStar(require("../component/BaseComponentGenerator"));
function composeWith(base, options) {
    Utils.compose(path.basename(__dirname), base, options);
}
exports.composeWith = composeWith;
function defineOptions(generator) {
    BaseComponent.defineOptions(generator, 'Web part');
}
exports.defineOptions = defineOptions;
class WebpartGenerator extends BaseComponent.BaseComponentGenerator {
    constructor(args, options) {
        super(args, options);
        this.friendlyName = 'Web part';
        this.codeName = 'WebPart';
        this.allowedTemplates = ['react', 'minimal'];
        this.folderName = 'webparts';
        defineOptions(this);
    }
    initializing() {
        /* no-op */
    }
    async prompting() {
        this.ensureCorrectFolder();
        if (this.shouldExecute()) {
            await super.prompting();
        }
    }
    configuring() {
        if (this.shouldExecute()) {
            super.configuring();
            this.ensureDependencyGroup('webpart');
            this.serveJson.setUpForWebParts();
            const solutionName = YeomanConfiguration_1.YeomanConfiguration.libraryName;
            this.packageSolutionJson.addFeature(Utils.generateGuid(), `${solutionName} Feature`, `The feature that activates elements of the ${solutionName} solution.`, false);
        }
    }
    writing() {
        if (this.shouldExecute()) {
            super.writing(false);
            const dest = this._getOutputFolder(this.context.componentNameCamelCase);
            const template = this.context.template;
            this.copyTemplate(path.join(this.sourceRoot(), template), dest);
            if (template !== 'minimal') {
                this.copyTemplate(path.join(this.sourceRoot(), 'loc'), path.join(dest, 'loc'));
                this.copyTemplate(path.join(this.sourceRoot(), 'assets'), path.join(dest, 'assets'));
            }
            this.copyTemplate(path.join(this.sourceRoot(), 'base'), dest);
            this.fs.copy(this.templatePath('./images/outline.png'), this.destinationPath(`teams/${this.context.componentId}_outline.png`));
            this.fs.copy(this.templatePath('./images/color.png'), this.destinationPath(`teams/${this.context.componentId}_color.png`));
        }
    }
    install() {
        if (this.shouldExecute()) {
            super.install();
        }
    }
    end() {
        super.end();
    }
    shouldExecute() {
        return this.config.get('environment') === 'spo' && this.config.get('componentType') === 'webpart';
    }
}
exports.WebpartGenerator = WebpartGenerator;
// Yeoman expects exports of this type, which overrides the TS exports above
// Ensure that anything marked as export above is also listed here
module.exports = WebpartGenerator;
module.exports.defineOptions = defineOptions;
module.exports.composeWith = composeWith;
