// @microsoft/generator-sharepoint/lib/generators/formCustomizer/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FormCustomizerGenerator = exports.defineOptions = exports.composeWith = void 0;
const path = __importStar(require("path"));
const Utils = __importStar(require("../../common/utilities"));
const YeomanConfiguration_1 = require("../../common/YeomanConfiguration");
const BaseExtension = __importStar(require("../extension/BaseExtensionGenerator"));
function composeWith(base, options) {
    Utils.compose(path.basename(__dirname), base, options);
}
exports.composeWith = composeWith;
function defineOptions(generator) {
    BaseExtension.defineOptions(generator, 'Form Customizer');
}
exports.defineOptions = defineOptions;
class FormCustomizerGenerator extends BaseExtension.BaseExtensionGenerator {
    constructor(args, options) {
        super(args, options);
        this.friendlyName = 'Form Customizer';
        this.codeName = 'FormCustomizer';
        this.allowedTemplates = ['react', 'none'];
        defineOptions(this);
    }
    initializing() {
        /* no-op */
    }
    prompting() {
        this.ensureCorrectFolder();
        if (!this.shouldExecute()) {
            return Promise.resolve();
        }
        return super.prompting();
    }
    configuring() {
        if (this.shouldExecute()) {
            super.configuring();
            this.ensureDependencyGroup('formCustomizer');
            // New Form
            this.serveJson.addConfiguration(`${this.context.componentNameCamelCase}_NewForm`, {
                pageUrl: 'https://{tenantDomain}/_layouts/15/SPListForm.aspx',
                formCustomizer: {
                    componentId: this.context.componentId,
                    PageType: 8,
                    RootFolder: '/sites/mySite/Lists/MyList',
                    properties: {
                        sampleText: 'Value'
                    }
                }
            });
            // Edit Form
            this.serveJson.addConfiguration(`${this.context.componentNameCamelCase}_EditForm`, {
                pageUrl: 'https://{tenantDomain}/_layouts/15/SPListForm.aspx',
                formCustomizer: {
                    componentId: this.context.componentId,
                    PageType: 6,
                    RootFolder: '/sites/mySite/Lists/MyList',
                    ID: 1,
                    properties: {
                        sampleText: 'Value'
                    }
                }
            });
            // View Form
            this.serveJson.addConfiguration(`${this.context.componentNameCamelCase}_ViewForm`, {
                pageUrl: 'https://{tenantDomain}/_layouts/15/SPListForm.aspx',
                formCustomizer: {
                    componentId: this.context.componentId,
                    PageType: 4,
                    RootFolder: '/sites/mySite/Lists/MyList',
                    ID: 1,
                    properties: {
                        sampleText: 'Value'
                    }
                }
            });
        }
    }
    writing() {
        if (this.shouldExecute()) {
            super.writing(false);
            const dest = this._getOutputFolder(this.context.componentNameCamelCase);
            const template = this.context.template;
            this.copyTemplate(path.join(this.sourceRoot(), template), dest);
            if (template !== 'minimal') {
                this.copyTemplate(path.join(this.sourceRoot(), 'loc'), path.join(dest, 'loc'));
            }
            this.copyTemplate(path.join(this.sourceRoot(), 'base'), dest);
        }
    }
    install() {
        if (this.shouldExecute()) {
            super.install();
        }
    }
    end() {
        super.end();
    }
    addFeature() {
        if (this.shouldExecute() && !this.hasElementsXml()) {
            const solutionName = YeomanConfiguration_1.YeomanConfiguration.libraryName;
            this.packageSolutionJson.addFeature(Utils.generateGuid(), `${solutionName} Feature`, `The feature that activates elements of the ${solutionName} solution.`, false);
        }
    }
    includeClientSideInstances() {
        return false;
    }
    shouldExecute() {
        return (this.config.get('environment') === 'spo' &&
            this.config.get('componentType') === 'extension' &&
            this.config.get('extensionType') === 'FormCustomizer');
    }
}
exports.FormCustomizerGenerator = FormCustomizerGenerator;
// Yeoman expects exports of this type, which overrides the TS exports above
// Ensure that anything marked as export above is also listed here
module.exports = FormCustomizerGenerator;
module.exports.defineOptions = defineOptions;
module.exports.composeWith = composeWith;
